src/tools/vfs/vfs_main.c

```diff
@@ -10,9 +10,370 @@
+}
+
+static void vfs_usage()
+{
+    struct sockaddr_un sock_addr = {};
+
+    ucs_vfs_sock_get_address(&sock_addr);
+    printf("Usage:   ucx_vfs [options]  [action]\n");
+    printf("\n");
+    printf("Options:\n");
+    printf("  -d <dir>   Set parent directory for mount points (default: %s)\n",
+           g_opts.mountpoint_dir);
+    printf("  -o <opts>  Pass these mount options to mount.fuse\n");
+    printf("  -f         Do not daemonize; run in foreground\n");
+    printf("  -v         Enable verbose logging (requires -f)\n");
+    printf("  -l <path>  Set listening unix socket path (default: %s)\n",
+           sock_addr.sun_path);
+    printf("\n");
+    printf("Actions:\n");
+    printf("   start     Run the daemon and listen for connection from UCX\n");
+    printf("             If a daemon is already running, do nothing\n");
+    printf("             This is the default action.\n");
+    printf("   stop      Stop the running daemon\n");
+    printf("\n");
+}
+
+static int vfs_parse_args(int argc, char **argv)
 {
-    /* TODO implement VFS daemon logic */
+    const char *action_str;
+    int c, i;
+
+    while ((c = getopt(argc, argv, "d:o:vfl:h")) != -1) {
+        switch (c) {
+        case 'd':
+            g_opts.mountpoint_dir = optarg;
+            break;
+        case 'o':
+            g_opts.mount_opts = optarg;
+            break;
+        case 'v':
+            ++g_opts.verbose;
+            break;
+        case 'f':
+            g_opts.foreground = 1;
+            break;
+        case 'l':
+            g_opts.sock_path = optarg;
+            break;
+        case 'h':
+        default:
+            vfs_usage();
+            return -127;
+        }
+    }
+
+    if (g_opts.verbose && !g_opts.foreground) {
+        vfs_error("Option -v requires -f");
+        vfs_usage();
+        return -1;
+    }
+
+    if (optind < argc) {
+        action_str    = argv[optind];
+        g_opts.action = UCS_VFS_SOCK_ACTION_LAST;
+        for (i = 0; i < ucs_static_array_size(vfs_action_names); ++i) {
+            if ((vfs_action_names[i] != NULL) &&
+                !strcmp(action_str, vfs_action_names[i])) {
+                g_opts.action = i;
+            }
+        }
+        if (g_opts.action == UCS_VFS_SOCK_ACTION_LAST) {
+            vfs_error("invalid action '%s'", action_str);
+            vfs_usage();
+            return -1;
+        }
+        ++optind;
+    }
+
+    if (optind < argc) {
+        vfs_error("only one action can be specified");
+        vfs_usage();
+        return -1;
+    }
+
     return 0;
 }
+
+static int vfs_test_fuse()
+{
+    char *argv[] = {"-V", NULL};
+    return vfs_run_fusermount(argv);
+}
+
+int main(int argc, char **argv)
+{
+    int ret;
+
+    ret = vfs_parse_args(argc, argv);
+    if (ret < 0) {
+        return -1;
+    }
+
+    ret = vfs_test_fuse();
+    if (ret < 0) {
+        return -1;
+    }
+
+    ret = mkdir(g_opts.mountpoint_dir, S_IRWXU);
+    if ((ret < 0) && (errno != EEXIST)) {
```

Comment:
why checking for `EEXIST`?
what if `mkdir()` fails with `errno == EACCES`?

Reply:
EEXIST means we can reuse this directory because it's same user id
EACCESS is real failure, it means need to specify a different directory

Reply:
> EEXIST means we can reuse this directory because it's same user id

ah, I see
sorry, misread this part of code